Ods-pz-refactor

eatyourpeas opened the pull request:
### Overview
A large refactor with not everything working - especially view preference and filtering

Maint things are to:
1. Move the PDU list to the PaediatricDiabetesUnit model: this sees parent and PDU names and lead organisation ods code and name. There is one for each in the UK and Jersey. RCPCH is included. It is seeded in the s/start-dev and has a function in manage.py to seed.
2. Implements a new endpoint from RCPCH NHS Organisations API to pull a list of PDUS (including RCPCH) with their parent and lead organisations and trust/health boards. There is as well as a function to compare with the API list and update accordingly - in future this will be run on a schedule.
3. Refactors the session object now to store only the pz_code. This is used draw the lists for the selects in the forms and the `view_progress`. These are now pz code and lead organisation and parent names.
4. ods code is nolonger used anywhere in the app for filtering and the sibling organisations concept has been removed. The API endpoint is still live
5. forms (esp NPDAUser) have been much simplified and duplicated logic removed. The forms handle no saving or select instantiation logic - this is all handled or passed in from the view.

Implications
The view preference does update the session but now does not remain set at the correct PDU. 
The csv_upload validates appropriately but now adds an item that is valid in the invalid list

eatyourpeas committed:
implement pz codes with parent and primary organisation names

eatyourpeas committed:
adds org and parent names/ods_codes to PDU model, adds untested seeder to manage.py, adds function to update local PDU table from API (untested) and simple API call for new endpoint

eatyourpeas committed:
working seed function

eatyourpeas committed:
function to update local records with api: maintain_paediatric_diabetes_unit_records_against_rcpch_nhs_organisations_API. Logging not working but function works

eatyourpeas committed:
refactor superuser to user RCPCH PZ code, non RCPCH superusers to user KCH PZ code

eatyourpeas committed:
refactors in organisation_employer calls queries to paediatric diabetes unit fixing all tests

eatyourpeas committed:
new function to retrieve PDU names for view preference
add to npdauser listview

eatyourpeas committed:
remove ods from session, add pdu seeder before createsuperuser, remove ods from signals

eatyourpeas committed:
fix htmx calls in npda_user forms and update view to use PaediatricDiabetesUnit model and pz_code
remove redundant duplicate methods in npda_user form

eatyourpeas committed:
replace pz numbers with names in employers list in npda user form template

eatyourpeas committed:
fix user form to instantiate with choices from view

eatyourpeas committed:
instantiate update user form in updateview with pdu list

eatyourpeas committed:
remove ods and organisation_choices and pdu_choices from session
remove same from view_preference and patientlistview

eatyourpeas committed:
remove ods code from updatepatient

eatyourpeas committed:
remove some print statements

eatyourpeas committed:
remove ods_code from csv_upload. now has small error allocation error

eatyourpeas committed:
fixing merge conflicts

eatyourpeas committed:
move pdu_choices to session on login

eatyourpeas committed:
view preference fixed
introduces new htmx url to update session and view preference and pass to partial

mbarton commented on project/npda/general_functions/organisations_adapter.py:
```diff
@@ -82,3 +88,105 @@ def organisations_to_populate_select_field(request, user_instance=None):
+                .exclude(npdauser=request.user)
+                .order_by("organisation_name")
+                .annotate(
+                    paediatric_diabetes_unit_name=Concat(
```
I'm not normally too hung up on DRY code but I think perhaps a function to build each of the annotate arguments would be useful to highlight what the differences are between each one?

It's hard to eyeball each one individually and understand what it's doing differently to the others

eatyourpeas replied:
fair point. sorry sent this in possibly before it was finished. Not every select needs the same list of PDUs - that is what this trying to do. I will fix and let you know when ready to review.

mbarton commented on project/npda/general_functions/rcpch_nhs_organisations.py:
```diff
@@ -131,116 +135,102 @@ def get_all_nhs_organisations_affiliated_with_paediatric_diabetes_unit() -> (
+        response = requests.get(url=url, timeout=10)  # times out after 10 seconds
+        response.raise_for_status()
+
+        if seed:
```
I think this function is only called with `seed=True`. Should we remove the argument entirely and the code below this guard?

mbarton commented on project/npda/general_functions/rcpch_nhs_organisations.py:
```diff
@@ -131,116 +135,102 @@ def get_all_nhs_organisations_affiliated_with_paediatric_diabetes_unit() -> (
+                        f"{new_pdu.pz_code} ({new_pdu.organisation_name}) was updated."
+                    )
+                else:
+                    logger.info(
```
I think this will be quite noisy in practice so probably one to remove before merging

eatyourpeas replied:
this is unlikely to be run more than weekly. it is polling the API and comparing the local list with what is there and updating as necessary. For sure can remove some of the logging. At the moment for some reason the logging is not working anyway.

mbarton commented on project/npda/models/npda_user.py:
```diff
@@ -104,37 +96,29 @@ def create_superuser(self, email, password, **extra_fields):
-                        pz_code="PZ215",
-                    )
+                paediatric_diabetes_unit = PaediatricDiabetesUnit.objects.get(
+                    pz_code="PZ215",  # Superusers that are not RCPCH staff are affiliated with King's College Hospital
```
Would we have any superusers not attached to a PDU or the RCPCH?

eatyourpeas replied:
no. I don't think so. I the group 4 are rcpch employees only. clinicians can have rcpch permissions but would have the is_rcpch_audit_team_member as true, unless you want us to remove that.

mbarton commented on project/npda/models/paediatric_diabetes_unit.py:
```diff
@@ -10,15 +10,32 @@ class PaediatricDiabetesUnit(Model):
         max_length=10,
         help_text="Enter the organisation ODS code",
     )
+    organisation_name = CharField(
```
Could we call these `lead_organisation` for clarity?

eatyourpeas committed:
refactor organisation_name and organisation_ods_code to lead_organisation_name and lead_organisation_ods_code

eatyourpeas committed:
add htmx custom trigger to patient list

eatyourpeas committed:
add htmx custom trigger to refresh npda user table

eatyourpeas committed:
add htmx get custom trigger to submissions list

eatyourpeas committed:
add pdu details for patient create/update form

eatyourpeas committed:
refactor and tidy `paediatric_diabetes_units_to_populate_select_field` and remove deprecated function from `organisations_adapter`

eatyourpeas committed:
refactor organisations_adapter further
ensure filtered pdus in user form context
in createnpdauser form set selected pdu to users

eatyourpeas committed:
make `maintain_paediatric_diabetes_unit_records_against_rcpch_nhs_organisations_API` less noisy

eatyourpeas committed:
remove seed flag from `get_all_pz_codes_with_their_trust_and_primary_organisation`

eatyourpeas committed:
remove request from paediatric_diabetes_units_to_populate_select_field as parameter and replace with requesting_user

eatyourpeas reviewed (commented):
I have addressed all the issues above and now the drop down is working across the app. The only item you raised in the createsuperuser function I have left unchanged but happy to amend.

I now have 5 failing tests though, which are all centre on pytest login with the corresponding request not containing a user object. I think this is the way we set up a user in the factory and log them in but have spent some time on it and can't for the moment resolve.

I will chat with anchit to see if we can fix. just wondering whether we might merge this anyway now that we actually have some test users. Whatever you feel in any case

eatyourpeas committed:
remove request from create_session function to fix tests

eatyourpeas committed:
fix final test and bugs around session in view_preference

mbarton commented:
Seen on [STAGING](https://npda-staging.rcpch.tech) (merged by @eatyourpeas 5 minutes and 23 seconds ago) Please check your changes!
